wgpu-hal/src/vulkan/adapter.rs

```diff
@@ -1071,33 +1101,51 @@ impl super::Adapter {
         }
 
         log::debug!("Supported extensions: {:?}", supported_extensions);
-        supported_extensions
+
+        let physical_device_features =
+            PhysicalDeviceFeatures::from_extensions_and_requested_features(
+                self.phd_capabilities.properties.api_version,
+                &supported_extensions,
+                features,
+                self.downlevel_flags,
+                &self.private_caps,
+                uab_types,
+            );
+
+        (supported_extensions, physical_device_features, uab_types)
     }
 
-    /// `features` must be the same features used to create `enabled_extensions`.
-    pub fn physical_device_features(
+    /// # Safety
+    /// `vk_device` must be created from this adapter, using `features`, `limits`, `family_index`
+    /// and at least `required_device_extensions()`, `physical_device_features()` using the same
+    /// features and limits.
+    pub unsafe fn device_from_raw(
         &self,
-        enabled_extensions: &[&'static CStr],
+        vk_device: u64,
+        handle_is_owned: bool,
         features: wgt::Features,
-        uab_types: super::UpdateAfterBindTypes,
-    ) -> PhysicalDeviceFeatures {
-        PhysicalDeviceFeatures::from_extensions_and_requested_features(
-            self.phd_capabilities.properties.api_version,
-            enabled_extensions,
+        limits: &wgt::Limits,
+        family_index: u32,
+        queue_index: u32,
+    ) -> Result<crate::OpenDevice<super::Api>, crate::DeviceError> {
+        let device =
+            ash::Device::load(self.instance.raw.fp_v1_0(), vk::Device::from_raw(vk_device));
+
+        let (extensions, _, uab_types) = self.required_device_capabilities(features, limits);
+
+        self.device_from_raw_inner(
+            device,
+            handle_is_owned,
+            &extensions,
             features,
-            self.downlevel_flags,
-            &self.private_caps,
             uab_types,
+            family_index,
+            queue_index,
         )
     }
 
-    /// # Safety
-    ///
-    /// - `raw_device` must be created from this adapter.
-    /// - `raw_device` must be created using `family_index`, `enabled_extensions` and `physical_device_features()`
-    /// - `enabled_extensions` must be a superset of `required_device_extensions()`.
     #[allow(clippy::too_many_arguments)]
-    pub unsafe fn device_from_raw(
+    pub unsafe fn device_from_raw_inner(
```

Comment:
does this still need to be public? strange given the "_inner" suffix

Reply:
That was a mistake, thanks. I'm going to finish some tests on the weekend.